src/chain/auth.sw

```diff
@@ -1,41 +1,160 @@
+//! Functionality for determining who is calling a contract.
 
-use ::result::Result;
 use ::address::Address;
+use ::b512::B512;
+use ::chain::assert;
 use ::contract_id::ContractId;
+use ::result::Result;
+
 
 pub enum AuthError {
     ContextError: (),
 }
 
 pub enum Sender {
     Address: Address,
-    Id: ContractId,
+    ContractId: ContractId,
 }
 
-/// Returns `true` if the caller is external (ie: a script or predicate).
-// ref: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#gm-get-metadata
+/// Returns `true` if the caller is external (i.e. a script).
+/// ref: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/opcodes.md#gm-get-metadata
 pub fn caller_is_external() -> bool {
     asm(r1) {
         gm r1 i1;
         r1: bool
     }
 }
 
-/// Get the `Sender` (ie: `Address`| ContractId) from which a call was made.
-/// Returns a Result::Ok(Sender) or Result::Error.
-// NOTE: Currently only returns Result::Ok variant if the parent context is Internal.
+fn caller_contract_id() -> ContractId {
+    ~ContractId::from(asm(r1) {
+        gm r1 i2;
+        r1: b256
+    })
+}
+
+/// Get the `Sender` (i.e. `Address`| `ContractId`) from which a call was made.
+/// Returns a `Result::Ok(Sender)` or `Result::Err` is a sender cannot be determined.
+/// NOTE: Currently only returns `Result::Ok` variant if the parent context is internal.
 pub fn msg_sender() -> Result<Sender, AuthError> {
     if caller_is_external() {
-        // TODO: Add call to get_coins_owner() here when implemented,
-        Result::Err(AuthError::ContextError)
+        let address = get_coins_owner();
+        if let Result::Ok(inner_value) = address {
+            Result::Ok(Sender::Address(address))
+        } else {
+            Result::Err(AuthError::ContextError)
+        }
     } else {
-        // Get caller's contract ID
-        let id = ~ContractId::from(asm(r1) {
-            gm r1 i2;
-            r1: b256
+        // Get caller's `ContractId`
+        Result::Ok(Sender::ContractId(caller_contract_id()))
+    }
+}
+
+/// If the input's type is `InputCoin`, return the owner. Otherwise, error.
+fn get_input_owner(input_ptr: u32) -> Result<Sender> {
+        // get data offest by 1 word
+        let data_ptr = asm(buffer, ptr: input_ptr, data_ptr) {
+            move buffer sp;
+            cfei i8;
+            addi data_ptr input_ptr i8;
+            mcpi buffer data_ptr i8;
+            buffer: u8
+        };
+
+        let owner_addr = ~Address::from(asm(buffer, ptr: data_ptr, owner_ptr) {
+            move buffer sp;
+            cfei i8;
+            addi owner_ptr data_ptr i16;
+            mcpi buffer owner_ptr i32;
+            buffer: b256
         });
-        Result::Ok(Sender::Id(id))
+
+        Result::Ok(Sender::Address(owner_addr))
+
+}
+
+/// Get the owner of the inputs(of type `InputCoin`) to a TransactionScript,
+/// if they all share the same owner.
+pub fn get_coins_owner() -> Result<Sender> {
```

Comment:
Should this be `Result<Sender, AuthError>` ?

Reply:
Oh yeah, it definitely should.